RFCs/rfc-0008-wallet-provider-api.md

```diff
@@ -0,0 +1,387 @@
+    method: 'mina_chainId'
+}): Promise<string>;
+```
+
+#### mina_sign
+Returns a signed message.
+
+##### Parameters
+`message`: [Required] a string
+
+##### Returns
+A signed message.
+
+```typescript
+type SignedMessage = {
+    publicKey: string;
+    data: string;
+    signature: {
+        field: string;
+        scalar: string;
+    };
+};
+
+Provider.request({
+    method: 'mina_sign',
+    params: { message: "hi bob" }
+}): Promise<SignedMessage>;
+```
+
+#### mina_signFields
+Returns a signed message.
+
+##### Parameters
+`fields`: [Required] an array of numbers or strings
+
+##### Returns
+
+```typescript
+type SignedFieldsData = {
+    publicKey: string;
+    data: (string | number)[];
+    signature: {
+        field: string;
+        scalar: string;
+    };
+};
+
+Provider.request({
+    method: 'mina_signFields',
+    params: { fields: [1, 5, 9, 2] }
+}): Promise<SignedFieldsData>;
+```
+
+#### mina_signTransaction
+Returns a signed transaction.
+
+##### Parameters
+`transaction`: [Required] a well-formed transaction body.
+
+##### Returns
+A signed transaction body.
+
+```typescript
+Provider.request({
+    method: 'mina_signTransaction',
+    params: { transaction: tx}
+}): Promise<SignedTransaction>;
+```
+
+#### mina_sendTransaction
+Submits a signed transaction to the network.
+
+##### Parameters
+- `signedTransaction`: [Required] the signed transaction body
+- `transactionBody`: [Required] the transaction body
+- `transactionType`: [Required] the type of the transaction (e.g. `payment`, `delegation`, `zkapp`)
+
+```typescript
+type TransactionReceipt = {
+    hash: string
+};
+
+Provider.request({
+    method: 'mina_sendTransaction',
+    params: { signedTransaction: signedTx, transactionBody: txBody, transactionType: 'payment'}
+}): Promise<TransactionReceipt>
+```
+#### mina_getBalance
+Returns the balance of the the client's current account.
+
+##### Parameters
+None.
+
+##### Returns
+```typescript
+Provider,request({
+    method: 'mina_getBalance'
+}): Promise<number>
+```
+
+#### mina_createNullifier
+Returns a nullifier.
+
+##### Parameters
+`message`: [Required] an array of `bigint`s
+
+##### Returns
+```typescript
+type Nullifier = {
+    publicKey: Group;
+    public: {
+        nullifier: Group;
+        s: Field;
+    };
+    private: {
+        c: Field;
+        g_r: Group;
+        h_m_pk_r: Group;
+    };
+};
+
+Provider.request({
+    method: 'mina_createNullifier',
+    params: { message: [1n] }
+}): Promise<Nullifier>
+```
+
+
+## Appendix I: Consumer-Facing API Documentation
+
+### request
+
+Makes a Mina RPC method call.
+
+```typescript
+interface RequestArguments {
+  readonly method: string;
+  readonly params?: readonly unknown[] | object;
+}
+
+Provider.request(args: RequestArguments): Promise<unknown>;
+```
+
+The returned Promise resolves with the method's result or rejects with a [`ProviderRpcError`](#errors). For example:
+
+```javascript
+Provider.request({ method: 'mina_accounts' })
```

Comment:
After zkApp request, when the current wallet account is not connected to zkApp, there are two methods:
- return silently
- A popup will pop up requesting to link the account

Because some zkApps hope to get zkApp requests silently, so as to avoid forced pop-ups as soon as zkApp is opened.